app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/ToolbarUiController.kt

```diff
@@ -36,17 +45,50 @@ class ToolbarUiController(
 
     private var hasUserChangedURLSinceEditTextFocused = false
 
+    private lateinit var tooltip: PopupWindow
+    private lateinit var tooltipView: View
+
     fun onCreateView(layout: View, viewLifecycleOwner: LifecycleOwner, fragmentManager: FragmentManager) {
         val toolbarClickListener = ToolbarOnClickListener()
         layout.topNavContainer.forEach {
             it.nextFocusDownId = layout.navUrlInput.id
             if (it.isFocusable) it.setOnClickListener(toolbarClickListener)
+
+            it.setOnFocusChangeListener { _, hasFocus ->
+                if (hasFocus) showTooltip(it)
+                else tooltip.dismiss() // Hide the tooltip when the button is not focused
+            }
         }
 
+        val layoutInflater = layout.context.systemService<LayoutInflater>(Context.LAYOUT_INFLATER_SERVICE)
+        tooltipView = layoutInflater.inflate(R.layout.tooltip, null)
+        tooltip = PopupWindow(tooltipView, WRAP_CONTENT, WRAP_CONTENT, false)
+
         setupUrlInput(layout)
         initToolbar(layout, viewLifecycleOwner, fragmentManager)
     }
 
+    private fun showTooltip(navBarButton: View) {
+        tooltip.contentView.tooltip.text = navBarButton.contentDescription
+        tooltip.isClippingEnabled = false
+
+        // The measurement of the popup happens in onGlobalLayout. We need to update the position of the
+        // popup after this measurement has happened. Beforehand, the measuredWidth will be 0.
+        tooltipView.viewTreeObserver.addOnGlobalLayoutListener(
+            object : ViewTreeObserver.OnGlobalLayoutListener {
+                override fun onGlobalLayout() {
+                    tooltip.update(
+                        navBarButton,
+                        0 - (tooltip.contentView.measuredWidth - navBarButton.width) / 2,
+                        10, -1, -1
```

Comment:
Nit: break these last 2 args into separate lines too, I think.

Reply:
Ok, I wasn't sure about that. They looked so empty all on separate lines.